virttest/qemu_vm.py

```diff
@@ -1038,10 +1038,35 @@ def add_memorys(devices, params):
                 options.append("maxmem=%s" % mem_params["maxmem"])
                 if mem_params.get("slots"):
                     options.append("slots=%s" % mem_params["slots"])
+            sgx_epc_memdev_id = dict()
             for name in params.objects("mem_devs"):
                 dev = devices.memory_define_by_params(params, name)
+                for ele in dev:
+                    if isinstance(ele, qdevices.Memory) and \
+                            ele.params["backend"] == "memory-backend-epc":
+                        sgx_epc_memdev_id[name] = ele.params["id"]
                 devs.extend(dev)
+            sgx_epc_cmd = ""
+            if params.get("vm_sgx_epc_devs"):
```

Comment:
we can remove if clause, use 
for idx, ele in enumerate(params.objects("vm_sgx_epc_devs")):

Reply:
Thanks Zhenchao, for vm_sgx_epc_devs, it is also a required option, if mem_devs has epc object(dict sgx_epc_memdev_id not None), then vm_sgx_epc_devs should be required. But params.objects("vm_sgx_epc_devs") couldn't return an error if 'vm_sgx_epc_devs' not exist, so I suggested changing to:
if sgx_epc_memdev_id:
                cap_type = "sgx-epc"
                for idx, ele in enumerate(params["vm_sgx_epc_devs"].split()):
@zhencliu what do you think?

Reply:
'if sgx_epc_memdev_id' is not needed here, unless we make sure the epc devs defined in vm_sgx_epc_devs can exactly match the ones defined in mem_devs, it's a little complicated(not needed) to do the checks, so, just let VT fail it. Whatever the user set the params, VT can fail it if the settings are not correct. 

using split is OK for me. cc @YongxueHong for more comments

Reply:
Hi @zhencliu 
If 'vm_sgx_epc_devs' is None, then no sgx-epc device will be added to machine cmd, in this case, VT and qemu won't fail, guest can still boot up, but the result is there are zero epc sections reported in guest dmesg. This should be treated as a fault configuration of sgx, because sgx isn't enabled in guest, this is also not a negative case. So I think it should add check for 'vm_sgx_epc_devs'.
For checking epc devs defined in vm_sgx_epc_devs can exactly match the ones defined in mem_devs, just need to check the dict length of sgx_epc_memdev_id equals to length of vm_sgx_epc_devs.

Reply:
e.g.       
           if sgx_epc_memdev_id and len(sgx_epc_memdev_id) == len(params.objects(
                        "vm_sgx_epc_devs")):
                cap_type = "sgx-epc"
                for idx, ele in enumerate(params.objects("vm_sgx_epc_devs")):

Reply:
> Hi @zhencliu If 'vm_sgx_epc_devs' is None, 
Then it fails due to VT cannot get the required param (params['vm_sgx_epc_devs'].split)

Reply:
> e.g. if sgx_epc_memdev_id and len(sgx_epc_memdev_id) == len(params.objects( "vm_sgx_epc_devs")): cap_type = "sgx-epc" for idx, ele in enumerate(params.objects("vm_sgx_epc_devs")):

OK for me.

Reply:
I change my mind after discussing with zhenchao. Based on qemu behavior, sgx-epc maps to epc memory devices, but not vice versa. I agree to remove if checks, just 
for idx, ele in numerate(params.objects("vm_sgx_epc_devs")):
if dict 'sgx_epc_memdev_id' does not have the key 'vm_sgx_epc_devs' wants, the test will fail.

Reply:
Agreed. 
If vm_sgx_epc_devs is *not* defined, even if epc-backend device is defined in mem_devs, we do nothing for sgx;
If vm_sgx_epc_devs is defined(e.g. vm_sgx_epc_devs=sgx0), but vm_sgx_epc_memdev_sgx0 is not defined in mem_devs, we should raise an exception, e.g.  virt_vm.VMDeviceNotFoundError('Cannot find epc memory device for sgx0')

Reply:
Update commit, 
for idx, ele in enumerate(params.objects("vm_sgx_epc_devs")):
                sgx_epc_params = params.object_params(ele)
                sgx_epc_memdev_ele = sgx_epc_params["vm_sgx_epc_memdev"]
                if sgx_epc_memdev_ele not in sgx_epc_memdev_id.keys():
                    raise virt_vm.VMDeviceNotFoundError("Cannot find epc "
                                                        "memory device mapped"
                                                        " for sgx-epc device "
                                                        "%s, please check "
                                                        "your sgx cfg." % ele)